src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+  //    new_variant = <loop_decreases>;
+  //   ASSERT(new_variant < old_variant);
+  //   ASSUME(false);
+
+  goto_programt pre_loop_latch_instrs;
+
+  {
+    // Record that we entered the loop with `entered_loop = true`
+    pre_loop_latch_instrs.add(
+      goto_programt::make_assignment(entered_loop, true_exprt{}));
+  }
+
+  {
+    // Jump back to the step case to havoc the write set, assume the invariant,
+    // and execute an arbitrary iteration.
+    pre_loop_latch_instrs.add(goto_programt::make_goto(
+      step_case_target, in_base_case, loop_head_location));
+  }
+
+  {
+    // Because of the unconditional jump above the following code is only
+    // reachable in the step case. Generate the inductive invariant check
+    // `ASSERT(invariant)`.
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_invariant_step");
+    check_location.set_comment(
+      "Check invariant after step for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    code_assertt assertion{invariant};
+    assertion.add_source_location() = check_location;
+    converter.goto_convert(assertion, pre_loop_latch_instrs, language_mode);
+  }
+
+  {
+    // Generate assignments to store the multidimensional decreases clause's
+    // value after one iteration of the loop
+    if(!decreases_clauses.empty())
+    {
+      for(size_t i = 0; i < new_decreases_vars.size(); i++)
+      {
+        code_assignt new_decreases_assignment{
+          new_decreases_vars[i], decreases_clauses[i]};
+        new_decreases_assignment.add_source_location() = loop_head_location;
+        converter.goto_convert(
+          new_decreases_assignment, pre_loop_latch_instrs, language_mode);
+      }
+
+      // Generate assertion that the multidimensional decreases clause's value
+      // after the loop is lexicographically smaller than its initial value.
+      source_locationt check_location(loop_head_location);
+      check_location.set_property_class("loop_decreases");
+      check_location.set_comment(
+        "Check variant decreases after step for loop " +
+        id2string(check_location.get_function()) + "." +
+        std::to_string(cbmc_loop_id));
+      code_assertt monotonic_decreasing_assertion{
+        generate_lexicographic_less_than_check(
+          new_decreases_vars, old_decreases_vars)};
+      monotonic_decreasing_assertion.add_source_location() = check_location;
+      converter.goto_convert(
+        monotonic_decreasing_assertion, pre_loop_latch_instrs, language_mode);
+
+      // Discard the temporary variables that store decreases clause's value
+      for(size_t i = 0; i < old_decreases_vars.size(); i++)
+      {
+        pre_loop_latch_instrs.add(
+          goto_programt::make_dead(old_decreases_vars[i], loop_head_location));
+        pre_loop_latch_instrs.add(
+          goto_programt::make_dead(new_decreases_vars[i], loop_head_location));
+      }
```

Comment:
@qinheping can you confirm: 

Exit blocks declares  DEAD all instrumentation variables that were DECL in the pre-header node and are alive on loop exit.

The DECL for `new_decreases_var` is inserted right before the latch instruction. Since the latch instruction is guaranteed to be an unconditional GOTO head (hence non-exiting) and since we turn it into an assume(false),  it is impossible to exit the loop with `new_decreases_var` alive. Hence we should only declare them DEAD in the loop body, but not in exit blocks.

Since we turn the latch instruction into assume(false), I guess we could say that all other instrumentation variables should also be declared DEAD here, but since progress stops this does not really matter.

Reply:
They should all be marked out-of-scope in the exit block. I will fix it.

Reply:
Exit blocks are reached only when exiting the loop before checking the loop step so I don't think the `new_decreases_vars` should be mentioned in the exit blocks. It is impossible to reach an exit block after executing the DECLs for the post-step decreases clause variables.